File: benchmarks/c/threaded_cpu_map.c
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int worker_id;
    int workers;
    int items;
    int rounds;
    int64_t checksum;
} WorkerArgs;

static int env_int(const char *name, int fallback) {
    const char *raw = getenv(name);
    char *end = NULL;
    long value;
    if (raw == NULL || raw[0] == '\0') {
        return fallback;
    }
    value = strtol(raw, &end, 10);
    if (end == raw || *end != '\0' || value <= 0 || value > 2147483647L) {
        return fallback;
    }
    return (int)value;
}

static int64_t heavy_work(int64_t seed, int rounds) {
    int64_t acc = seed + 1;
    int i;
    for (i = 0; i < rounds; i++) {
        acc = (acc * 1103515245LL + 12345LL + i + seed) % 2147483647LL;
    }
    return acc % 1000003LL;
}

static void *worker_sum(void *ptr) {
    WorkerArgs *args = (WorkerArgs *)ptr;
    int64_t checksum = 0;
    int i;
    for (i = args->worker_id; i < args->items; i += args->workers) {
        checksum += heavy_work(i, args->rounds);
    }
    args->checksum = checksum;
    return NULL;
}

int main(void) {
    int workers = env_int("BENCH_THREADS", 4);
    int items = env_int("BENCH_ITEMS", 10000);
    int rounds = env_int("BENCH_ROUNDS", 1000);
    pthread_t *threads = (pthread_t *)calloc((size_t)workers, sizeof(pthread_t));
    WorkerArgs *args = (WorkerArgs *)calloc((size_t)workers, sizeof(WorkerArgs));
    int64_t checksum = 0;
    int i;

    if (threads == NULL || args == NULL) {
        free(threads);
        free(args);
        return 1;
    }

    for (i = 0; i < workers; i++) {
        args[i].worker_id = i;
        args[i].workers = workers;
        args[i].items = items;
        args[i].rounds = rounds;
        if (pthread_create(&threads[i], NULL, worker_sum, &args[i]) != 0) {
            free(threads);
            free(args);
            return 1;
        }
    }

    for (i = 0; i < workers; i++) {
        pthread_join(threads[i], NULL);
        checksum += args[i].checksum;
    }

    printf("checksum: %lld\n", (long long)checksum);
    free(threads);
    free(args);
    return 0;
}

File: benchmarks/python/threaded_cpu_map.py
import os
import sys
import sysconfig
from concurrent.futures import ThreadPoolExecutor


def require_free_threaded():
    if sys.version_info < (3, 14):
        raise SystemExit("Python concurrency benchmarks require Python 3.14+")
    if str(sysconfig.get_config_var("Py_GIL_DISABLED")) != "1":
        raise SystemExit("Python concurrency benchmarks require a free-threaded build")
    is_gil_enabled = getattr(sys, "_is_gil_enabled", None)
    if callable(is_gil_enabled) and is_gil_enabled():
        raise SystemExit("Python concurrency benchmarks require the GIL disabled")


def env_int(name, fallback):
    try:
        value = int(os.environ.get(name, ""))
    except ValueError:
        return fallback
    return value if value > 0 else fallback


def heavy_work(seed, rounds):
    acc = seed + 1
    for i in range(rounds):
        acc = (acc * 1103515245 + 12345 + i + seed) % 2147483647
    return acc % 1000003


def worker_sum(worker_id, workers, items, rounds):
    checksum = 0
    for i in range(worker_id, items, workers):
        checksum += heavy_work(i, rounds)
    return checksum


def main():
    require_free_threaded()
    workers = env_int("BENCH_THREADS", 4)
    items = env_int("BENCH_ITEMS", 10000)
    rounds = env_int("BENCH_ROUNDS", 1000)
    with ThreadPoolExecutor(max_workers=workers) as pool:
        futures = [
            pool.submit(worker_sum, worker_id, workers, items, rounds)
            for worker_id in range(workers)
        ]
        checksum = sum(future.result() for future in futures)
    print(f"checksum: {checksum}")


if __name__ == "__main__":
    main()
